### ISSUE TITLE
Data missing from 'Multi value traits' CSV output?

### ISSUE THREAD

**Issue by AdamDimech** [danforthcenter/plantcv#412]
I have run a pipeline over a [folder of images](https://plantcv.readthedocs.io/en/latest/pipeline_parallel/#running-plantcv-workflows-over-a-flat-directory-of-images) and generated output in a JSON file, which was then [converted to two CSV files](https://plantcv.readthedocs.io/en/latest/tools/#plantcv-utilities). The multi-value-traits output is cryptic and I need some help to understand what it is, given the output.

I amended my pipeline workflow code via `pcv.outputs.add_observation` to enable me to add [custom output](https://plantcv.readthedocs.io/en/latest/outputs/) to the JSON file (and thus single value sheet):

`filestring = filename`
`uid = filename[:-22] #Remove file extensions and time stamp`
`pcv.outputs.add_observation(variable='Filename', trait='Name of file used in imaging', method='Filename derived', scale='Name', datatype=int, value=filestring, label='Filename')`
`pcv.outputs.add_observation(variable='Plant_ID', trait='Name of accession used in imaging', method='Filename derived', scale='Name', datatype=int, value=uid, label='Unique_ID')`

This works well in the single value sheet but there is nothing in the multi-value sheet to link the output to an individual ID.

The output in the multi-value-traits sheet is in the following format (all other columns contain rows with `NA` values:

<table class="tg">
  <tr>
    <th class="tg-73oq">trait</th>
    <th class="tg-73oq">value</th>
    <th class="tg-73oq">label</th>
  </tr>
  <tr>
    <td class="tg-73oq">blue_frequencies</td>
    <td class="tg-73oq">4</td>
    <td class="tg-73oq">0</td>
  </tr>
  <tr>
    <td class="tg-73oq">blue_frequencies</td>
    <td class="tg-73oq">1</td>
    <td class="tg-73oq">1</td>
  </tr>
  <tr>
    <td class="tg-73oq">blue_frequencies</td>
    <td class="tg-73oq">5</td>
    <td class="tg-73oq">2</td>
  </tr>
  <tr>
    <td class="tg-73oq">blue_frequencies</td>
    <td class="tg-73oq">6</td>
    <td class="tg-73oq">3</td>
  </tr>
  <tr>
    <td class="tg-73oq">blue_frequencies</td>
    <td class="tg-73oq">6</td>
    <td class="tg-73oq">4</td>
  </tr>
  <tr>
    <td class="tg-73oq">blue_frequencies</td>
    <td class="tg-73oq">9</td>
    <td class="tg-73oq">5</td>
  </tr>
  <tr>
    <td class="tg-73oq">blue_frequencies</td>
    <td class="tg-73oq">22</td>
    <td class="tg-73oq">6</td>
  </tr>
</table>

I ran an analysis of 3,500 images, yet I only have 255 values of `blue_frequencies` (and every other parameter in the sheet), which suggests that either this is an aggrigate result for all 3,500 images or more likely, a result for just one of the images analysed. Is the data missing? Is this a programming bug? I have the histogram output [as image files](https://github.com/danforthcenter/plantcv/issues/391#issuecomment-504448763), but not as numerical data for each image.

I wonder whether a `pcv.outputs.add_metadata` might be useful here?

Clarification would be appreciated.

**Comment by nfahlgren** [comment]
Hi @AdamDimech,

I suspect your multi-value-traits file is incomplete/truncated. There was a bug in `pcv.analyze_color` in the `hue` observation that caused `plantcv-utils.py json2csv` to crash prior to PlantCV `v3.4.1`. I think once we can figure out what is going on in issue #401 that part of this issue will be fixed.

The format of the single-value-traits table is one row per image (with some caveats). All the single-value traits are separate columns in this table.

It's not really possible to fit the multi-value traits into the above kind of format, so instead the multi-value-trait table is formatted in the R/tidyverse "long" format. Each row is a trait, a value, and a label. In your example above, there is one trait, the value is pixel counts, and the label is the frequency table bin ID. In this case the blue histogram/frequency table has 256 values for blue color channel intensity values of 0-255. In R I could then do something like this to plot the blue channel histograms:

```R
library(ggplot2)

data = read.table(file = "test-multi-value-traits.csv", sep = ",", header = TRUE, 
            na.strings = "NA")

ggplot(data[data$trait == "blue_frequencies",], aes(x = label, y = value, 
            color = plantbarcode)) +
            geom_point()
```

When you run `plantcv-workflow.py` it tries to parse the filenames to extract metadata. The metadata are included in the output JSON file and the `json2csv` converter includes the metadata columns in both the single- and multi-value trait tables, which allows you to merge them as necessary.

It sounds like you your setup though your script is extracting the metadata, and you are right, while you can add an observation, there's not currently a way to add metadata through the same mechanism. Because the metadata you are adding goes into the `observations` data, `json2csv` is not treating it as metadata, so it's not included in both tables.

Bottom line is I think you are right, we need a `pcv.outputs.add_metadata` method too to fix this issue.

**Comment by AdamDimech** [comment]
Thanks for the explanation @nfahlgren. Much appreciated.

**Comment by HaleySchuhl** [comment]
I'm going to close this issue since we now have one open requesting to add metadata
